File: api/src/main/java/br/com/tradugil/traducao/Tokenizador.java
package br.com.tradugil.traducao;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Quebra o texto em palavras e monta os candidatos a gíria.
 *
 * <p>Como {@code Normalizador}, tem gêmeas em {@code packages/core-ts} e no
 * módulo Android: o destaque das gírias precisa cair exatamente nas mesmas
 * posições no site, na extensão e no aplicativo.</p>
 */
public final class Tokenizador {

    /**
     * Sequências de letras ou dígitos. A pontuação fica de fora de propósito:
     * ela é ruído para a busca, e o intervalo devolvido continua apontando
     * para a posição correta no texto original mesmo assim.
     */
    private static final Pattern PALAVRA = Pattern.compile("[\\p{IsAlphabetic}\\p{IsDigit}]+");

    /**
     * Maior verbete de várias palavras que vale procurar ("no cap", "dar
     * ruim"). Três é o teto porque a cada palavra a mais o número de
     * candidatos cresce junto com o custo da consulta, e verbetes de quatro
     * palavras são raros o bastante para não pagarem esse preço.
     */
    public static final int MAXIMO_DE_PALAVRAS = 3;

    private Tokenizador() {
    }

    /**
     * Um trecho do texto original e sua forma de busca.
     *
     * @param inicio índice do primeiro caractere, inclusivo
     * @param fim    índice logo após o último caractere, exclusivo — mesma
     *               convenção de {@code String.substring}, para o cliente
     *               poder destacar o trecho sem ajustar nada
     */
    public record Trecho(String original, String normalizado, int inicio, int fim) {
    }

    public static List<Trecho> tokenizar(String texto) {
        List<Trecho> palavras = new ArrayList<>();
        if (texto == null || texto.isBlank()) {
            return palavras;
        }
        Matcher m = PALAVRA.matcher(texto);
        while (m.find()) {
            String bruto = m.group();
            palavras.add(new Trecho(bruto, Normalizador.normalizar(bruto), m.start(), m.end()));
        }
        return palavras;
    }

    /**
     * Monta os candidatos, das sequências mais longas para as mais curtas.
     *
     * <p>A ordem é o que garante o casamento guloso mais adiante: quando o
     * texto tem "dar ruim", o candidato de duas palavras aparece antes de
     * "dar" e de "ruim" sozinhos, e vence. Sem isso, o usuário veria "ruim"
     * explicado como adjetivo comum em vez da expressão que ele leu.</p>
     */
    public static List<Trecho> candidatos(List<Trecho> palavras) {
        List<Trecho> candidatos = new ArrayList<>();
        for (int tamanho = MAXIMO_DE_PALAVRAS; tamanho >= 1; tamanho--) {
            for (int i = 0; i + tamanho <= palavras.size(); i++) {
                List<Trecho> janela = palavras.subList(i, i + tamanho);
                String normalizado = janela.stream()
                        .map(Trecho::normalizado)
                        .reduce((a, b) -> a + " " + b)
                        .orElse("");
                if (normalizado.isBlank()) {
                    continue;
                }
                int inicio = janela.get(0).inicio();
                int fim = janela.get(janela.size() - 1).fim();
                candidatos.add(new Trecho(normalizado, normalizado, inicio, fim));
            }
        }
        return candidatos;
    }
}

File: android/app/src/main/java/br/com/tradugiria/android/dominio/Tokenizador.kt
package br.com.tradugiria.android.dominio

/**
 * Quebra o texto em palavras e monta os candidatos a gíria.
 *
 * PORTE de `Tokenizador.java` e `tokenizador.ts`. O destaque das gírias
 * precisa cair exatamente nas mesmas posições no site, na extensão e aqui.
 */
object Tokenizador {

    /**
     * Sequências de letras ou dígitos. A pontuação fica de fora de propósito:
     * é ruído para a busca, e o intervalo devolvido continua apontando para a
     * posição correta no texto original mesmo assim.
     */
    private val PALAVRA = Regex("[\\p{IsAlphabetic}\\p{IsDigit}]+")

    /**
     * Maior verbete de várias palavras que vale procurar ("no cap", "dar
     * ruim"). Três é o teto porque a cada palavra a mais o número de
     * candidatos cresce junto com o custo da consulta, e verbetes de quatro
     * palavras são raros o bastante para não pagarem esse preço.
     */
    const val MAXIMO_DE_PALAVRAS = 3

    /**
     * Um trecho do texto original e sua forma de busca.
     *
     * [fim] é exclusivo — mesma convenção de `substring`, para a interface
     * marcar o trecho sem ajustar nada.
     */
    data class Trecho(
        val original: String,
        val normalizado: String,
        val inicio: Int,
        val fim: Int,
    )

    fun tokenizar(texto: String?): List<Trecho> {
        if (texto.isNullOrBlank()) return emptyList()

        return PALAVRA.findAll(texto).map { encontro ->
            Trecho(
                original = encontro.value,
                normalizado = Normalizador.normalizar(encontro.value),
                inicio = encontro.range.first,
                // `range.last` é inclusivo no Kotlin; somar 1 é o que mantém
                // a mesma convenção semiaberta das outras duas plataformas.
                fim = encontro.range.last + 1,
            )
        }.toList()
    }

    /**
     * Monta os candidatos, das sequências mais longas para as mais curtas.
     *
     * A ordem é o que garante o casamento guloso: quando o texto tem "dar
     * ruim", o candidato de duas palavras aparece antes de "dar" e "ruim"
     * sozinhos, e vence. Sem isso o usuário veria "ruim" explicado como
     * adjetivo comum, em vez da expressão que ele leu.
     */
    fun candidatos(palavras: List<Trecho>): List<Trecho> {
        val lista = mutableListOf<Trecho>()
        for (tamanho in MAXIMO_DE_PALAVRAS downTo 1) {
            for (i in 0..(palavras.size - tamanho)) {
                val janela = palavras.subList(i, i + tamanho)
                val normalizado = janela.joinToString(" ") { it.normalizado }
                if (normalizado.isBlank()) continue

                lista += Trecho(
                    original = normalizado,
                    normalizado = normalizado,
                    inicio = janela.first().inicio,
                    fim = janela.last().fim,
                )
            }
        }
        return lista
    }
}
